network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/pixels/WireguardHandshakeMonitor.kt

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2023 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.networkprotection.impl.pixels
+
+import android.content.Context
+import android.net.ConnectivityManager
+import android.net.ConnectivityManager.NetworkCallback
+import android.net.Network
+import android.net.NetworkCapabilities
+import android.net.NetworkRequest
+import com.duckduckgo.app.global.DispatcherProvider
+import com.duckduckgo.app.utils.ConflatedJob
+import com.duckduckgo.di.scopes.VpnScope
+import com.duckduckgo.mobile.android.vpn.service.VpnServiceCallbacks
+import com.duckduckgo.mobile.android.vpn.state.VpnStateMonitor.VpnStopReason
+import com.duckduckgo.networkprotection.api.NetworkProtectionState
+import com.duckduckgo.networkprotection.impl.WgProtocol
+import com.squareup.anvil.annotations.ContributesMultibinding
+import java.util.concurrent.atomic.AtomicBoolean
+import javax.inject.Inject
+import kotlin.time.Duration.Companion.minutes
+import kotlin.time.Duration.Companion.seconds
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import logcat.LogPriority.WARN
+import logcat.logcat
+import org.threeten.bp.Instant
+
+@ContributesMultibinding(VpnScope::class)
+class WireguardHandshakeMonitor @Inject constructor(
+    private val wgProtocol: WgProtocol,
+    private val pixels: NetworkProtectionPixels,
+    private val dispatcherProvider: DispatcherProvider,
+    private val networkProtectionState: NetworkProtectionState,
+    private val currentNetworkState: CurrentNetworkState,
+) : VpnServiceCallbacks {
+
+    private val job = ConflatedJob()
+    private val failureReported = AtomicBoolean(false)
+
+    override fun onVpnStarted(coroutineScope: CoroutineScope) {
+        job += coroutineScope.launch {
+            failureReported.set(false)
```

Comment:
Why are we reseting failureReported in onVpnStarted only?

Reply:
Is not only reset here, see the calls to `ailureReported.getAndSet`.

where else would you reset it?

Reply:
In the scenario where user has apptp and netp enabled. User disables netp after experiencing tunnel failure (`failureReported` will be set to true). If the user re-enables Netp (handshake diff will be less than threshold), do we expect `reportTunnelFailureRecovered` to be emitted?

Reply:
or maybe just at line `82`?